File: examples/cpp/livedata/src/LiveDataExample.cpp
#include <iostream>
#include <iomanip>
#include <thread>
#include <chrono>

#include "icsneo/icsneocpp.h"
#include "icsneo/communication/message/livedatamessage.h"
#include "icsneo/communication/livedata.h"

int main() {
	// Print version
	std::cout << "Running libicsneo " << icsneo::GetVersion() << std::endl;
	std::cout << "\nFinding devices... " << std::flush;
	auto devices = icsneo::FindAllDevices(); // This is type std::vector<std::shared_ptr<icsneo::Device>>
	std::cout << "OK, " << devices.size() << " device" << (devices.size() == 1 ? "" : "s") << " found" << std::endl;

	// List off the devices
	for(auto& device : devices)
		std::cout << '\t' << device->describe() << " @ Handle " << device->getNeoDevice().handle << std::endl;
	std::cout << std::endl;

	for(auto& device : devices) {
		std::cout << "Connecting to " << device->describe() << "... ";
		bool ret = device->open();
		if(!ret) { // Failed to open
			std::cout << "FAIL" << std::endl;
			std::cout << icsneo::GetLastError() << std::endl << std::endl;
			continue;
		}
		std::cout << "OK" << std::endl;

		// Create a subscription message for the GPS signals
		std::cout << "\tSending a live data subscribe command... ";
		auto msg = std::make_shared<icsneo::LiveDataCommandMessage>();
		msg->appendSignalArg(icsneo::LiveDataValueType::GPS_LATITUDE);
		msg->appendSignalArg(icsneo::LiveDataValueType::GPS_LONGITUDE);
		msg->appendSignalArg(icsneo::LiveDataValueType::GPS_ACCURACY);
		msg->appendSignalArg(icsneo::LiveDataValueType::DAQ_OVERRIDE);
		msg->appendSignalArg(icsneo::LiveDataValueType::MANUAL_TRIGGER);
		msg->cmd = icsneo::LiveDataCommand::SUBSCRIBE;
		msg->handle = icsneo::LiveDataUtil::getNewHandle();
		msg->updatePeriod = std::chrono::milliseconds(100);
		msg->expirationTime = std::chrono::milliseconds(0);
		// Transmit the subscription message
		ret = device->subscribeLiveData(msg);
		std::cout << (ret ? "OK" : "FAIL") << std::endl;

		// Register a handler that uses the data after it arrives every ~100ms
		std::cout << "\tStreaming messages for 3 seconds... " << std::endl << std::endl;
		auto filter = std::make_shared<icsneo::MessageFilter>(icsneo::Message::Type::LiveData);
		auto handler = device->addMessageCallback(std::make_shared<icsneo::MessageCallback>(filter, [&msg](std::shared_ptr<icsneo::Message> message) {
			auto ldMsg = std::dynamic_pointer_cast<icsneo::LiveDataMessage>(message);
			switch(ldMsg->cmd) {
				case icsneo::LiveDataCommand::STATUS: {
					auto msg2 = std::dynamic_pointer_cast<icsneo::LiveDataStatusMessage>(message);
					std::cout << "[Handle] " << ldMsg->handle << std::endl;
					std::cout << "[Requested Command] " << msg2->requestedCommand << std::endl;
					std::cout << "[Status] " << msg2->status << std::endl << std::endl;
					break;
				}
				case icsneo::LiveDataCommand::RESPONSE: {
					auto valueMsg = std::dynamic_pointer_cast<icsneo::LiveDataValueMessage>(message);
					if((valueMsg->handle == msg->handle) && (valueMsg->values.size() == msg->args.size()))
					{
						std::cout << "[Handle] " << msg->handle << std::endl;
						std::cout << "[Values] " << valueMsg->numArgs << std::endl;
						for(uint32_t i = 0; i < valueMsg->numArgs; ++i) {
							std::cout << "[" << msg->args[i]->valueType << "] ";
							auto scaledValue = icsneo::LiveDataUtil::liveDataValueToDouble(*valueMsg->values[i]);
							std::cout << scaledValue << std::endl;
						}
						std::cout << std::endl;
					}
					break;
				}
				default: // Ignoring other commands
					break;
			}
		}));
		// Run handler for three seconds to observe the signal data
		std::this_thread::sleep_for(std::chrono::seconds(3));
		auto setValMsg = std::make_shared<icsneo::LiveDataSetValueMessage>();
		setValMsg->cmd = icsneo::LiveDataCommand::SET_VALUE;
		setValMsg->handle = msg->handle;
		bool onOff = true;
		for (unsigned int i = 0; i < 10; ++i)
		{
			setValMsg->args.clear();
			setValMsg->values.clear();
			double val;
			if (onOff)
			{
				val = 1;
			}
			else
			{
				val = 0;
			}
			onOff = !onOff;
			icsneo::LiveDataValue ldValue;
			if (icsneo::LiveDataUtil::liveDataDoubleToValue(ldValue, val) < 0)
			{
				break;
			}
			setValMsg->appendSetValue(icsneo::LiveDataValueType::DAQ_OVERRIDE, ldValue);
			device->setValueLiveData(setValMsg);
			// Test setting values
			std::this_thread::sleep_for(std::chrono::seconds(3));
		}
		// Unsubscribe from the GPS signals and run handler for one more second
		// Unsubscription only requires a valid in-use handle, in this case from our previous subscription
		ret = device->unsubscribeLiveData(msg->handle);
		// The handler should no longer print values
		std::this_thread::sleep_for(std::chrono::seconds(1));
		device->removeMessageCallback(handler);
		std::cout << "\tDisconnecting... ";
		ret = device->close();
		std::cout << (ret ? "OK\n" : "FAIL\n") << std::endl;
	}
	return 0;
}

File: examples/python/livedata/livedata_example.py
"""
LiveData subscription and monitoring example using icsneopy library.

"""

import icsneopy
import time
from datetime import timedelta


def livedata_example():
	"""Subscribe to and monitor LiveData signals."""
	devices = icsneopy.find_all_devices()
	if not devices:
		raise RuntimeError("No devices found")

	device = devices[0]
	print(f"Using device: {device}")

	try:
		if not device.open():
			raise RuntimeError("Failed to open device")

		if not device.go_online():
			raise RuntimeError("Failed to go online")

		device.enable_message_polling()

		# Create subscription message
		msg = icsneopy.LiveDataCommandMessage()
		msg.handle = icsneopy.get_new_handle()
		msg.cmd = icsneopy.LiveDataCommand.SUBSCRIBE
		msg.update_period = timedelta(milliseconds=500)
		msg.expiration_time = timedelta(milliseconds=0)

		# Subscribe to various LiveData signals
		msg.append_signal_arg(icsneopy.LiveDataValueType.GPS_LATITUDE)
		msg.append_signal_arg(icsneopy.LiveDataValueType.GPS_LONGITUDE)
		msg.append_signal_arg(icsneopy.LiveDataValueType.GPS_ACCURACY)
		msg.append_signal_arg(icsneopy.LiveDataValueType.DAQ_ENABLE)
		msg.append_signal_arg(icsneopy.LiveDataValueType.MANUAL_TRIGGER)
		msg.append_signal_arg(icsneopy.LiveDataValueType.TIME_SINCE_MSG)

		print("\nSubscribing to LiveData signals...")
		if not device.subscribe_live_data(msg):
			raise RuntimeError(f"Subscription failed: {icsneopy.get_last_error()}")

		print("Subscription successful")
		print("\nMonitoring LiveData for 5 seconds...")

		response_count = 0
		start_time = time.time()

		while time.time() - start_time < 5:
			result = device.get_messages()
			messages = result[0] if isinstance(result, tuple) else result

			for m in messages:
				if isinstance(m, icsneopy.LiveDataStatusMessage):
					if m.handle == msg.handle:
						print(f"\n[Status] Command: {m.requested_command}, Status: {m.status}")

				elif isinstance(m, icsneopy.LiveDataValueMessage):
					if m.handle == msg.handle:
						response_count += 1
						print(f"\n[Response #{response_count}]")
						signal_names = ["GPS_LAT", "GPS_LON", "GPS_ACC", 
									   "DAQ_EN", "MAN_TRIG", "TIME_SINCE"]
						for idx, val in enumerate(m.values):
							value = icsneopy.livedata_value_to_double(val)
							name = signal_names[idx] if idx < len(signal_names) else f"Signal_{idx}"
							print(f"  {name:12s}: {value:10.2f}")

			time.sleep(0.1)

		print(f"\nReceived {response_count} response messages")

		# Demonstrate setting values
		print("\nSetting custom values...")
		set_msg = icsneopy.LiveDataSetValueMessage()
		set_msg.handle = icsneopy.get_new_handle()
		set_msg.cmd = icsneopy.LiveDataCommand.SET_VALUE

		# Set DAQ_ENABLE
		value = icsneopy.livedata_double_to_value(1.0)
		if value:
			set_msg.append_set_value(icsneopy.LiveDataValueType.DAQ_ENABLE, value)

		# Set MANUAL_TRIGGER
		value = icsneopy.livedata_double_to_value(1.0)
		if value:
			set_msg.append_set_value(icsneopy.LiveDataValueType.MANUAL_TRIGGER, value)

		if device.set_value_live_data(set_msg):
			print("Values set successfully")
			time.sleep(0.5)

			# Check the results
			result = device.get_messages()
			messages = result[0] if isinstance(result, tuple) else result
			for m in messages:
				if isinstance(m, icsneopy.LiveDataStatusMessage):
					if m.handle == set_msg.handle:
						print(f" Set status: {m.status}")

		# Keep device awake by resetting TIME_SINCE_MSG
		print("\nResetting TIME_SINCE_MSG to keep device awake...")
		reset_msg = icsneopy.LiveDataSetValueMessage()
		reset_msg.handle = icsneopy.get_new_handle()
		reset_msg.cmd = icsneopy.LiveDataCommand.SET_VALUE
		
		value = icsneopy.livedata_double_to_value(0.0)
		if value:
			reset_msg.append_set_value(icsneopy.LiveDataValueType.TIME_SINCE_MSG, value)
			if device.set_value_live_data(reset_msg):
				print("TIME_SINCE_MSG reset to 0")

		# Unsubscribe
		print("\nUnsubscribing...")
		if device.unsubscribe_live_data(msg.handle):
			print("Unsubscribed successfully")

	finally:
		device.close()
		print("\nDevice closed")


if __name__ == "__main__":
	livedata_example()

